### PR TITLE
BZ2023838: Adds better explanation of SCCs

### PR DESCRIPTION
**Pull request by darshan-nagaraj**
This PR is taken over from #38803. I have improvised the content to meet our writing standards.

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2023838

OCE version: 4.6 +

Doc Preview link: https://deploy-preview-39717--osdocs.netlify.app/openshift-enterprise/latest/authentication/managing-security-context-constraints.html#security-context-constraints-about_configuring-internal-oauth

@dbaker-rh and @xingxingxia PTAL and let me know if you have any comments.

### CODE REVIEW

**Comment by xingxingxia** [comment]
@y4sht , could you help review?

**Comment by dbaker-rh** [comment]
Looks good to me.

**Review by y4sht — CHANGES_REQUESTED** [review]

File: modules/security-context-constraints-about.adoc [hunk 1]
```diff
@@ -109,12 +110,12 @@ The `privileged` SCC allows:
 
 [NOTE]
 ====
-Setting `privileged: true` in the pod specification does not select the `privileged` SCC. Setting `privileged: true` in the pod specification matches on the `allowPrivilegedContainer` field of an SCC.
+Setting `privileged: true` in the pod specification does not select the `privileged` SCC. However, setting `privileged: true` in the pod specification considers only those SCCs that have a match on their `allowPrivilegedContainer` field, of which the `privileged` SCC is one.
```
[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: sure. I will make this change.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: In my opinion, mentioning ``However, setting `privileged: true`` is necessary because it's a key-value pair and also user may get confused if we mention this in singular third person pronoun. 

Also if you see the previous version of the documentation, it starts with ``setting `privileged: true``.
image

Let me know your thoughts on this.

[the remark this answers is not in the document]
Comment (y4sht) [reply modules/security-context-constraints-about.adoc:113]: "However" doesn't seem to justify the importance of the words that follow,  if that makes any sense, could be considered just a nit.

[the remark this answers is not in the document]
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:113]: Instead of nit picking this wording, we can rephrase it entirely to see if it reads more clearly.

For example:

Setting privileged: true in the pod specification does not necessarily select the privileged SCC.  The most appropriate SCC (of which the privileged SCC is one by default) which has "allowPrivilegedContainer: true" will be chosen if the user has permission to use it.

[the remark this answers is not in the document]
Comment (y4sht) [reply modules/security-context-constraints-about.adoc:113]: > Instead of nit picking this wording, we can rephrase it entirely to see if it reads more clearly.
> 
> For example:
> 
> Setting privileged: true in the pod specification does not necessarily select the privileged SCC. The most appropriate SCC (of which the privileged SCC is one by default) which has "allowPrivilegedContainer: true" will be chosen if the user has permission to use it.

👍 Just a minor edit

Setting `privileged: true` in the pod specification does not necessarily select the `privileged` SCC. The most appropriate SCC containing "allowPrivilegedContainer: true" will be selected if the user has permission to use it.

[the remark this answers is not in the document]
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:113]: Yes!  Your reword of my rephrase is ALMOST perfect, and very close to my original intent.

The problem is that if you take a regular workload and just throw "privileged: true" at it, you're actually (from experience) more likely to get admitted through the "node-exporter" SCC, and not privilged SCC at all.

[the remark this answers is not in the document]
Comment (y4sht) [reply modules/security-context-constraints-about.adoc:113]: Agreed, how about

Setting `privileged: true` in the pod specification does not necessarily select the `privileged` SCC. The most appropriate SCC containing `"allowPrivilegedContainer: true"` (out of which the `privileged` SCC is one by default) will be selected if the user has permission to use it.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: @y4sht We usually avoid putting content in parenthesis because users tend to ignore the content in parenthesis. There are exceptions like when expanding abbreviations, SI units.

[the remark this answers is not in the document]
Comment (l4zy0n3) [reply modules/security-context-constraints-about.adoc:113]: @xingxingxia suggested
Setting `privileged: true` in the pod specification does not necessarily select the `privileged` SCC. The most appropriate SCC which has `"allowPrivilegedContainer: true"` will be chosen if the user has permission to use it. The prioritization section of this documentation determines which one is the most appropriate.

[the remark this answers is not in the document]
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:113]: I think this phrasing is great.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: @l4zy0n3 
I didn't understand this:
`The prioritization section of this documentation determines which one is the most appropriate. ?`
Is this an explanation or you want to include this as well ?

[the remark this answers is not in the document]
Comment (y4sht) [reply modules/security-context-constraints-about.adoc:113]: Yes, it would be great if we could add this explaination, it makes the phrasing much clearer.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: @y4sht The prioritization section of this documentation - - which section you're referring to ? I wanna give cross-reference to that section.

[the remark this answers is not in the document]
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:113]: It's here https://deploy-preview-39717--osdocs.netlify.app/openshift-enterprise/latest/authentication/managing-security-context-constraints.html#scc-prioritization_configuring-internal-oauth

[the remark this answers is not in the document]
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:113]: (and it has a REALLY weird anchor tag name.  Should just be #scc-prioritization I think, but if it's being linked to from elsewhere we don't want to break those links)

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: @dbaker-rh i feel it's better not to touch anchor tags. As you mentioned it might break if it's reference somewhere else.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: @y4sht @dbaker-rh @xingxingxia I can't cross reference **Security context constraints prioritization** section because it's a part of  **About security context constraints** module. In our doc repo we have defined rule that we can't reference reference module in a module.

Let me give a more context on why we can't cross-reference module in a module -- explained by one of our DPM .
Having xref in a module, it breaks the contract of a module. It is no longer a module if it taking you to somewhere else, and that was one of the reasons this rule was put into place.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: @y4sht @dbaker-rh expect cross reference, I have made the other changes. Please let me know if there's any other way we can put this up.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc]: @y4sht PTAL and let me know your feedback.

[the remark this answers is not in the document]
Comment (y4sht) [reply modules/security-context-constraints-about.adoc]: @darshan-nagaraj what about
The SCC which has `allowPrivilegedContainer: true` and has the highest prioritization will be chosen if the user has the permission to use it.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc]: @y4sht Yes, makes sense. I will make the changes.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc]: @y4sht Changes updated. PTAL and if it's okay for you give me QA ack. Thanks

File: modules/security-context-constraints-about.adoc [hunk 2]
```diff
@@ -109,12 +110,12 @@ The `privileged` SCC allows:
 
 [NOTE]
 ====
-Setting `privileged: true` in the pod specification does not select the `privileged` SCC. Setting `privileged: true` in the pod specification matches on the `allowPrivilegedContainer` field of an SCC.
+Setting `privileged: true` in the pod specification does not necessarily select the `privileged` SCC. The SCC which has `allowPrivilegedContainer: true` and has the highest prioritization will be chosen if the user has the permission to use it.
```
Comment (ctauchen) [inline modules/security-context-constraints-about.adoc:113]:
```suggestion
Setting `privileged: true` in the pod specification does not necessarily select the `privileged` SCC. The SCC that has `allowPrivilegedContainer: true` and has the highest prioritization will be chosen if the user has the permissions to use it.
```

File: modules/security-context-constraints-about.adoc [hunk 3]
```diff
@@ -109,12 +110,12 @@ The `privileged` SCC allows:
...
 ====
 endif::[]
 
 |`restricted`
-|Denies access to all host features and requires pods to be run with a UID, and SELinux context that are allocated to the namespace.  This is the most restrictive SCC and it is used by default for authenticated users.
+|Denies access to all host features and requires pods to be run with a UID, and SELinux context that are allocated to the namespace.  This is the most restrictive default SCC and it is used by default for authenticated users.
```
Comment (ctauchen) [inline modules/security-context-constraints-about.adoc:118]: I suspect that "default" isn't the right word. We can't have multiple defaults for the same thing, right?

Perhaps you mean something closer to standard, prebuilt, preset?

File: modules/security-context-constraints-about.adoc [hunk 4]
```diff
@@ -124,6 +125,12 @@ The `restricted` SCC:
 * Requires that a pod is run with a pre-allocated MCS label
 * Allows pods to use any FSGroup
 * Allows pods to use any supplemental group
+
+[NOTE]
+====
+The `restricted` SCC still configures `allowPrivilegeEscalation: true`, which disables the `no_new_privs` protection. If you want to prevent this in your cluster, you can create a custom SCC with the flag disabled and configure the access to only appropriate users in place of `restricted`. For more details about flag, see link:https://www.kernel.org/doc/html/latest/userspace-api/no_new_privs.html[The Linux Kernel documentation].
```
Comment (ctauchen) [inline modules/security-context-constraints-about.adoc:131]: I get confused from "If you want to prevent this in your cluster". What is "this"? Configuring the first key? Disabling the second? Both? 

It's also not clear exactly what "the flag" refers to. Can you make all of this explicit?

**Review by ctauchen — COMMENTED** [review]
Hi! Thanks for this. Several line edits and questions below.

File: modules/security-context-constraints-about.adoc [hunk 5]
```diff
@@ -9,7 +9,8 @@ Similar to the way that RBAC resources control user access, administrators can u
 
 Security context constraints allow an administrator to control:
 
-* Whether a pod can run privileged containers
+* Whether a pod can run privileged containers (allowPrivilegedContainer)
+* Whether a pod is constrained by `no_new_privs` (allowPrivilegeEscalation)
```
[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:13]: @dbaker-rh could you please pitch in and clarify @ctauchen questions.

File: modules/security-context-constraints-about.adoc [hunk 6]
```diff
@@ -54,7 +55,7 @@ ifndef::openshift-dedicated[]
 
 [WARNING]
 ====
-This SCC allows host access to namespaces, file systems, and PIDS. It should only be used by trusted pods. Grant with caution.
+This SCC allows host access to namespaces, file systems, and PIDs. It should only be used by trusted pods. Grant with caution.
```
[the remark this answers is not in the document]
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:58]: My change here was simply to drop capitalization on the S.

If we're going to reword this better I think the whole sentence could be constructed better - PIDs is a namespace, but it 's somewhat confusing to refer to this as having "host access to ... the pid namespace".

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:58]: @ctauchen I checked our repo and couldn't find anywhere PIDs spelled out. It is mentioned as PID.

File: modules/security-context-constraints-about.adoc (hunk 3 above)
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:118]: @dbaker-rh could you please pitch in and clarify @ctauchen question.

File: modules/security-context-constraints-about.adoc (hunk 4 above)
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:131]: @dbaker-rh need your suggestion here.

File: modules/security-context-constraints-about.adoc (hunk 2 above)
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:113]: @ctauchen I will fix it. Thanks

File: modules/security-context-constraints-about.adoc (hunk 3 above)
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:118]: Yes.  This phrase, however inaccurate, is used in our own docs already: https://docs.openshift.com/container-platform/4.9/authentication/managing-security-context-constraints.html#default-sccs_configuring-internal-oauth

"The cluster contains several **default security context constraints** (SCCs) as described in the table below."

The intent of my change was to separate "most restrictive possible" (as is implied) from "most restrictive if you've not made any changes" (which is true).

To take a bash at rephrasing this:

>> ... This is the most restrictive SCC provided by a new installation and will be used by default for authenticated users.


If that gets too convoluted then leaving in it's mildly inaccurate current state might be better than explaining it to death.

File: modules/security-context-constraints-about.adoc (hunk 4 above)
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:131]: edit: ... which disables the "no_new_privs" protection flag.

edit: ... if you want the additional protection from disabling no_new_privs, you can create ...

File: modules/security-context-constraints-about.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbaker-rh) [reply modules/security-context-constraints-about.adoc:13]: What I added here was clarification ... before it just says "... can run privileged containers", but there are TWO separate and distinct flags within the SCC that have the word "privilege" in the name.

This change suggested to remove confusion over privilege vs privileged.  If it's done the opposite then I would suggest rewording to something more generic to cover both flags and then qualify them later.

For instance:

* What privileges the containers within the pod may be granted.

[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:13]: @dbaker-rh and @ctauchen I have restructured the bulleted lists. Does it makes sense? 

- Whether a pod can run privileged containers with `allowPrivilegedContainer` flag.

- Whether a `no_new_privs` pod is constrained with `allowPrivilegeEscalation` flag.

[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/security-context-constraints-about.adoc:13]: I think this is an improvement from what's there, for sure. Would add "the" before the flag name.

File: modules/security-context-constraints-about.adoc (hunk 3 above)
Comment (ctauchen) [reply modules/security-context-constraints-about.adoc:118]: Ok, we'll go with the usage from the docs. I like that phrasing: 

This is the most restrictive SCC provided by a new installation and will be used by default for authenticated users.

File: modules/security-context-constraints-about.adoc [hunk 7]
```diff
@@ -124,6 +125,12 @@ The `restricted` SCC:
 * Requires that a pod is run with a pre-allocated MCS label
 * Allows pods to use any FSGroup
 * Allows pods to use any supplemental group
+
+[NOTE]
+====
+The `restricted` SCC still configures `allowPrivilegeEscalation: true`, which disables the `no_new_privs` protection flag. If you want an additional protection from disabling `no_new_privs`, you can create a custom SCC with the flag disabled and configure the access to only appropriate users in place of `restricted`. For more details about flag, see link:https://www.kernel.org/doc/html/latest/userspace-api/no_new_privs.html[The Linux Kernel documentation].
```
Comment (ctauchen) [inline modules/security-context-constraints-about.adoc:131]: I'm not confident I understand this. Particularly the second sentence. Perhaps this is worth making clearer?

If you want an additional protection from disabling `no_new_privs`,(meaning: if you want most of what you get in `restricted`, but with `no_new_privs` enabled?) you can create a custom SCC with the flag (which flag?) disabled and configure the access to only appropriate users _in place of `restricted`_(what is `restricted` in place of? users?)

File: modules/security-context-constraints-about.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:13]: @ctauchen Fixed!

File: modules/security-context-constraints-about.adoc (hunk 3 above)
Comment (darshan-nagaraj) [reply modules/security-context-constraints-about.adoc:118]: @ctauchen Done!

File: modules/security-context-constraints-about.adoc (hunk 7 above)
Comment (ctauchen) [reply modules/security-context-constraints-about.adoc:131]: Perhaps something like this:

````
[NOTE]
====
In the `restricted` SCC, `allowPrivilegeEscalation` is set to `true`, which disables the `no_new_privs` protection flag. If you want to enable `no_new_privs` and use an SCC that is similar to the `restricted` SCC, you can create a custom SCC that uses the same configuration but with `allowPrivilegeEscalation` is set to `true`.
````
Since the note isn't about `no_new_privs`, but rather about how to enable it with a restricted security context, I suggest we leave out the external link.

Comment (ctauchen) [reply modules/security-context-constraints-about.adoc:131]: @dbaker-rh This version is clearer, but may be inaccurate. Can you confirm this is right? Thanks!
